File: python/secant.py
import sympy as sp


def secant(x, f, x0, x1, n, tol):
    fx0 = f.subs(x, x0)

    inicial = x0
    if fx0 == 0:
        print(float(x0), "Es raíz")
    else:
        fx1 = f.subs(x, x1)
        df = sp.diff(f, x)
        i = 0;
        er = tol + 1
        den = fx1 - fx0
        while er > tol and fx1 != 0 and den != 0 and i < n:
            print("i:{:03d} x: {:.10f} fx: {:.10f} error abs: {:.10f} den: {:.10f}\n"
                  .format(i + 1, float(x1), float(fx1), float(er),float(den)), end="")
            x2 = x1 - fx1 * (x1 - x0) / den
            er = abs(x2 - x1)
            x0 = x1
            fx0 = fx1
            x1 = x2
            fx1 = f.subs(x, x1)
            den = fx1 - fx0
            i = i + 1
        print("i:{:03d} x: {:.10f} fx: {:.10f} error abs: {:.10f} \n"
              .format(i + 1, float(x1), float(fx1), float(er)), end="")
        if fx1 == 0:
            print(float(x1), "Es raíz")
        elif er < tol:
            print(float(x1), ' es aproximación a una raíz con tolerancia ', tol)
        elif den == 0:
            print(float(x1), ' es una posible raíz multiple')
        else:
            print('El método fracasó con ', i, ' iteraciones.')
        sp.plot(f, (x, inicial, x1 + n * tol))

File: matlab/equationsOneVariable/secant.m

function [] = secant(f,x0,x1,n,tol)

syms x;
inicial=x0;
sp = '%.10f';
spa='%.10f %.10f %.10f %.10f\n';
disp (['i ' 'x ' 'fx ' 'errorabs ']);
fx0=double(subs(f,x0));
if fx0==0
    disp([num2str(x0,sp) ' is root.']);
else
    fx1=double(subs(f,x1));
    df=matlabFunction( diff(f(x)));
    i=0;
    er=tol+1;
    den=fx1-fx0;
    resp=[0 double(x0) double(fx0) nan];
    fprintf( spa, resp.' ) ;
    while er>tol && fx1~=0 && den~=0 && i<n
        if i==0
            er=nan;
        end
        resp=[i+1 double(x1) double(fx1) double(er)];
        fprintf( spa, resp.' ) ;
        x2=x1-fx1*(x1-x0)/den;
        er=abs(x2-x1);
        x0=x1;
        fx0=fx1;
        x1=x2;
        fx1=double(subs(f,x1));
        den=fx1-fx0;
        i=i+1;
    end
    resp=[i+1 double(x1) double(fx1) double(er)];
    fprintf( spa, resp.' ) ;
    if fx1==0
        disp([num2str(x1,sp) ' is root.']);
    elseif er<tol
        disp([num2str(x1,sp) 'is an approximation with tolerance ' num2str(tol,sp)]);

    elseif den==0
         disp('There might be a multiple root');
    else
        disp(['The method failed with ' num2str(i) ' iterations.']);
    end
end

ejex = linspace(inicial, x1+100*tol);
figure
plot(ejex, double(subs(f,ejex)))
grid
end
